pkg/api/mlflow/dao/repositories/param.go

```diff
@@ -31,8 +32,50 @@ func NewParamRepository(db *gorm.DB) *ParamRepository {
 
 // CreateBatch creates []models.Param entities in batch.
 func (r ParamRepository) CreateBatch(ctx context.Context, batchSize int, params []models.Param) error {
-	if err := r.db.CreateInBatches(params, batchSize).Error; err != nil {
+	var err error
+	err = r.db.CreateInBatches(params, batchSize).Error
+	if err != nil {
+		// remove duplicate rows and try again
+		params, err1 := r.removeExactMatches(ctx, params)
+		if err1 != nil {
+			return eris.Wrap(err1, "error removing exact matches after: " + err.Error())
+		}
+		err = r.db.CreateInBatches(params, batchSize).Error
+	}
+
+	if err != nil {
 		return eris.Wrap(err, "error creating params in batch")
 	}
 	return nil
 }
+
+// removeExactMatches will return a new slice of params which excludes exact matches
+func (r ParamRepository) removeExactMatches(ctx context.Context, params []models.Param) ([]models.Param, error) {
+	var keys, foundKeys []string
+	paramMap := map[string]models.Param{}
+	paramsToReturn := []models.Param{}
+	for _, param := range params {
+		key := fmt.Sprintf("%v-%v-%v", param.RunID, param.Key, param.Value)
+		keys = append(keys, key)
+		paramMap[key] = param
+	}
+
+	tx := r.db.Raw(`
+           select run_uuid || '-' || key || '-' || value
+           from params
+           where run_uuid || '-' || key || '-' || value in ?`, keys).
+		Find(&foundKeys)
+	if tx.Error != nil {
+		return paramsToReturn, eris.Wrap(tx.Error, "problem selecting existing params")
+	}
+
+	for _, foundKey := range foundKeys {
```

Comment:
ohh, we have 3 cycles. are we 100% sure about that?

Reply:
no sure if there's a way to avoid without some "set operations" library or the like, which also range over collections as implementation

Reply:
with `hashes` we can immediately remove all duplicates and re create new ones.

Reply:
well the issue is we have to subtract a set of keys from the hash in this implementation -- but maybe I'm not seeing it